New blog post

DanielHara opened the pull request:
Dear visitor, your opinion is wanted!
Could you please grant me a quick read and report about our impressions?
You may be as harsh as you'd like.

The repository for trying at is https://github.com/DanielHara/unit-testing-react

cgebken commented:
Hey @DanielHara :) Thanks for writing this post. I had a quick first look. Just a few thoughts that came to my mind: 

- We have a convention to only write one sentence per line. This way it's easier for us to review and keep an overview.
- We try not to start a post with a headline. It would be great if you could come up with a few introductory sentences before the first headline.
- You use unit testing and unit-testing. It would be better to use a consistent spelling.

I'll ask Aroso to check the rest. But I think the general structure and topic is fine :)

One last thing regarding the sample repository: Great idea :) Would it be possible to make it an ePages related one so that we can maintain it and it's not all in your responsibility?

DanielHara commented:
+> Hey @DanielHara :) Thanks for writing this post. I had a quick first look. Just a few thoughts that came to my mind:
> 
> * We have a convention to only write one sentence per line. This way it's easier for us to review and keep an overview.
> * We try not to start a post with a headline. It would be great if you could come up with a few introductory sentences before the first headline.
> * You use unit testing and unit-testing. It would be better to use a consistent spelling.
> * If you search for your header image, I recommend to use https://pixabay.com/ or https://www.pexels.com/ so that we don't have any copyright issues.
> 
> I'll ask Aroso to check the rest. But I think the general structure and topic is fine :)
> 
> One last thing regarding the sample repository: Great idea :) Would it be possible to make it an ePages related one so that we can maintain it and it's not all in your responsibility?

Hi @cgebken , thank you very much for your comments!
I have addressed the issues you have just mentioned, and created the try-out repo: https://github.com/DanielHara/unit-testing-react

`ePages` can simply copy this repository to an own repository of its and I'll then delete mine accordingly.

amahaiuddin commented on _posts/2020/2020-02-11-unit-testing-in-react.md:
```diff
@@ -0,0 +1,251 @@
+On the one hand, in unit testing, you test each one of your functions and components individually. On the other, in integration testing, you test
+how all those work along with each other to deliver all the fancy features you offer your users.
+
+As Adrià Fontcuberta pointed out at his remarkable talk at dotJS in Paris (check out this [blog post](/blog/events/dotjs-2019-in-paris-from-the-perspective-of-a-frontend-designer/){:target="_blank"} about the conference, as well as the [video](https://www.dotconferences.com/2019/12/adria-fontcuberta-the-pragmatic-front-end-tester){:target="_blank"} of the actual talk), tests give predictability a ...
```
I would recommend to add "conference" after "dotJS".

amahaiuddin commented on _posts/2020/2020-02-11-unit-testing-in-react.md:
```diff
@@ -0,0 +1,251 @@
+
+## Getting started
+
+One of our most recent features at our storefront was a "Load more" button in the product pages. We could draft here a simplified version of this
```
I think it must be "... on the product pages".

amahaiuddin commented on _posts/2020/2020-02-11-unit-testing-in-react.md:
```diff
@@ -0,0 +1,251 @@
+## Getting started
+
+One of our most recent features at our storefront was a "Load more" button in the product pages. We could draft here a simplified version of this
+feature, and carry on to writing tests for it! This will help us to understand how tests work and how to ensure our tests are making our code
```
Sugesstion: "... and how to ensure that ..."

DanielHara replied:
In this case, I guess it is indifferent. I find the variant without "that" more direct and closer to spoken informal dialogue.

amahaiuddin commented on _posts/2020/2020-02-11-unit-testing-in-react.md:
```diff
@@ -0,0 +1,251 @@
+feature, and carry on to writing tests for it! This will help us to understand how tests work and how to ensure our tests are making our code
+reliable and predictable.
+
+We could start out with a simple component which has a button which can fetch (via an API) a fixed amount of, say, 1 product, and stores them
```
Suggestion to avoid "which" twice in a sentence: "... a button that ..."

DanielHara replied:
> Is "1 product" an example? If so, it's maybe better to write for example: "e.g."
Yes, the use of the word "say" already highlights that the number of "1" is an arbitrary suggestion.

amahaiuddin commented on _posts/2020/2020-02-11-unit-testing-in-react.md:
````diff
@@ -0,0 +1,251 @@
+export default ProductList;
+```
+
+Before going on to how to test it, I must say I've prepared a sample repo for you to try out all the code in this post.
````
Maybe you can write the exact word, instead of "it": "Before going on to how to test **it**, ..."

DanielHara replied:
I didn't understand this comment. What do you mean?

amahaiuddin commented on _posts/2020/2020-02-11-unit-testing-in-react.md:
```diff
@@ -0,0 +1,251 @@
+
+## Let's test it!
+
+Now comes the question, how would you test the component `ProductList`?
```
Suggestion: " Now we come to the question, ..."

DanielHara replied:
Why? It's the same thing.

ingmarh commented on _posts/2020/2020-02-11-unit-testing-in-react.md:
````diff
@@ -0,0 +1,251 @@
+```javascript
+const defaultFetchProducts = (page) => fetch(`http://localhost/products?page=${page}`)
+  .then(response => response.json())
+  .catch((error) => console.log(error))
````
For brevity in this blog post, I suggest to leave out the `.catch()`.

DanielHara replied:
done

ingmarh commented on _posts/2020/2020-02-11-unit-testing-in-react.md:
```diff
@@ -0,0 +1,251 @@
+  .then(response => response.json())
+  .catch((error) => console.log(error))
+
+const ProductList = ({ fetchProducts=defaultFetchProducts }) => {
```
```suggestion
const ProductList = ({ fetchProducts = defaultFetchProducts }) => {
```

DanielHara replied:
done - I linted it quickly with standard

ingmarh commented on _posts/2020/2020-02-11-unit-testing-in-react.md:
```diff
@@ -0,0 +1,251 @@
+  const [page, setPage] = useState(1)
+  const loadProducts = async () => {
+    const response = await fetchProducts(page)
+    if(response) {
```
```suggestion

    if (response) {
```

DanielHara replied:
done

ingmarh commented on _posts/2020/2020-02-11-unit-testing-in-react.md:
```diff
@@ -0,0 +1,251 @@
+  expect(fetchProducts).toHaveBeenCalledTimes(1)
+  expect(fetchProducts).toHaveBeenCalledWith(1)
+
+  //click the second time
```
For consistency:

```suggestion
  // click the second time
```

DanielHara replied:
done

ingmarh commented on _posts/2020/2020-02-11-unit-testing-in-react.md:
```diff
@@ -0,0 +1,251 @@
+This test can already be taken seriously! It checks whether we called `fetchProducts` with the right arguments and whether we
+use the result of these calls in a meaningful way.
+
+On the repo, this approach is on branch `mocking-fetch-products`.
```
I would avoid the short notation "repo" in favour of writing the full word.

```suggestion
In the example repository, you can find this approach on the branch `mocking-fetch-products`.
```

DanielHara replied:
done

ingmarh commented on _posts/2020/2020-02-11-unit-testing-in-react.md:
```diff
@@ -0,0 +1,251 @@
+
+The former approach has, nevertheless, its drawbacks. We rely completely on the mock of `fetchProducts`. How can we know if it would
+hit the right API endpoints?
+There's were the awesome `nock` library comes along. You can also mock the HTTP requests! Calling `nock('http://localhost')` will
```
Same as written in the other annotation: I think we shouldn't use the package names, but instead the official spelling of the names according to the corresponding libraries (and possibly link them for the first occurrence in the text).

```suggestion
That's were the awesome Nock library comes along. You can also mock the HTTP requests! Calling `nock('http://localhost')` will
```

DanielHara replied:
done - but here I really mean "There"

ingmarh commented on _posts/2020/2020-02-11-unit-testing-in-react.md:
```diff
@@ -0,0 +1,251 @@
+The former approach has, nevertheless, its drawbacks. We rely completely on the mock of `fetchProducts`. How can we know if it would
+hit the right API endpoints?
+There's were the awesome `nock` library comes along. You can also mock the HTTP requests! Calling `nock('http://localhost')` will
+mock any requests made to `http://localhost` _inside_ our test! This way we also test the right HTTP requests are being made
```
```suggestion
mock any requests made to `http://localhost` _inside_ our test! This way, we also test that the right HTTP requests are being made
```

DanielHara replied:
done

ingmarh commented on _posts/2020/2020-02-11-unit-testing-in-react.md:
```diff
@@ -0,0 +1,251 @@
+  fireEvent.click(loadMoreButton)
+  await act(() => Bluebird.delay(500))
+
+  //expect two products
```
```suggestion
  // expect two products
```

DanielHara replied:
done

ingmarh commented on _posts/2020/2020-02-11-unit-testing-in-react.md:
````diff
@@ -0,0 +1,251 @@
+})
+```
+
+You can try this out on the `master` branch of the repo!
````
_See the other annotation_.

```suggestion
You can try this out on the `master` branch of the example repository.
```

Also, I think these hint paragraphs to the example repository would look good in a info box, too (see other annotation).

DanielHara replied:
done

ingmarh commented on _posts/2020/2020-02-11-unit-testing-in-react.md:
```diff
@@ -0,0 +1,251 @@
+`scope.done()` call, which is an awesome feature of `nock`! It asserts that all of the mocked API endpoint calls have been hit!
+And, furthermore, this tests the component's feature independently of its implementation (so called 'black-box testing').
+It does not matter for the test, for instance, how the `page` parameter is stored in the component. This makes it a lot easier
+when the need for _refactoring_ comes.
```
```suggestion
when the component needs to be refactored at a later point.
```

DanielHara replied:
done

ingmarh replied:
But, it's the same as before. Is this intended?

ingmarh commented on _posts/2020/2020-02-11-unit-testing-in-react.md:
```diff
@@ -0,0 +1,251 @@
+## Conclusion
+
+Nowadays, there is absolutely no reason why you should not unit test your code. With so many great tools out there, the effort
+of writing tests absolutely pays off by dramatically decreasing the quantity of bugs and getting you covered when refactoring.  🎉
```
👍

DanielHara commented on _data/authors.yml:
```diff
@@ -148,6 +148,12 @@
+- id: dhara
+  name: Daniel
+  full_name: Daniel Hara
+  description: is a Javascript Enginner and a classical music and opera aficionado who is passionate about testing.
```
Done!

DanielHara commented on _posts/2020/2020-02-11-unit-testing-in-react.md:
```diff
@@ -0,0 +1,251 @@
+
+## Getting started
+
+One of our most recent features at our storefront was a "Load more" button in the product pages. We could draft here a simplified version of this
```
done

DanielHara replied:
done

DanielHara commented on _posts/2020/2020-02-11-unit-testing-in-react.md:
```diff
@@ -0,0 +1,251 @@
+
+const fetchProducts = (page) => fetch(`http://localhost/products?page=${page}`)
+  .then(response => response.json())
+  .catch((error) => console.log(error))
```
done

DanielHara commented on _posts/2020/2020-02-11-unit-testing-in-react.md:
```diff
@@ -0,0 +1,251 @@
+    <div>{product.name}</div>
+    <div>{product.price}</div>
+  </div>
+)
```
done

DanielHara commented on _posts/2020/2020-02-11-unit-testing-in-react.md:
```diff
@@ -0,0 +1,251 @@
+
+Now comes the question, how would you test the component `ProductList`?
+
+Here at ePages we use extensively the libraries `react-testing-library` and `jest` to write our unit tests.
```
done

DanielHara commented on _posts/2020/2020-02-11-unit-testing-in-react.md:
````diff
@@ -0,0 +1,251 @@
+export default ProductList;
+```
+
+Before going on to how to test it, I must say I've prepared a sample repo for you to try out all the code in this post.
````
done

DanielHara commented on _posts/2020/2020-02-11-unit-testing-in-react.md:
```diff
@@ -0,0 +1,251 @@
+       Load more!
+    </button>
+    {products.map(product => <Product {...product} />)}
+  </div>
```
Using ```jsx fixes it!

DanielHara commented on _posts/2020/2020-02-11-unit-testing-in-react.md:
```diff
@@ -0,0 +1,251 @@
+## Mock it
+
+`jest` offers you the possibility of mocking all kinds of functions, including their return values and assert how many times they were called,
+and with which arguments they were called. It also offers a great deal of synthatic sugar to make your tests look pretty and for you to admire them
```
done
